pkg/handlers/defaulthandler.go

```diff
@@ -0,0 +1,235 @@
+
+	switch archive := format.(type) {
+	case archiver.Decompressor:
+		// Decompress tha archive and feed the decompressed data back into the archive handler to extract any nested archives.
+		compReader, err := archive.OpenReader(arReader)
+		if err != nil {
+			return err
+		}
+
+		defer compReader.Close()
+
+		return h.openArchive(ctx, depth+1, compReader, archiveChan)
+	case archiver.Extractor:
+		return archive.Extract(logContext.WithValue(ctx, depthKey, depth+1), arReader, nil, h.extractorHandler(archiveChan))
+	default:
+		return fmt.Errorf("unknown archive type: %s", format.Name())
+	}
+}
+
+// extractorHandler creates a closure that handles individual files extracted by an archiver.
+// It logs the extraction, checks for cancellation, and decides whether to skip the file based on its name or type,
+// particularly for binary files if configured to skip. If the file is not skipped, it recursively calls openArchive
+// to handle nested archives or to continue processing based on the file's content and depth in the archive structure.
+func (h *DefaultHandler) extractorHandler(archiveChan chan []byte) func(context.Context, archiver.File) error {
+	return func(ctx context.Context, file archiver.File) error {
+		lCtx := logContext.WithValues(
+			logContext.AddLogger(ctx),
+			"filename", file.Name(),
+			"size", file.Size(),
+		)
+		lCtx.Logger().V(5).Info("Handling extracted file.")
+
+		if common.IsDone(ctx) {
+			return ctx.Err()
+		}
+
+		depth := 0
+		if ctxDepth, ok := ctx.Value(depthKey).(int); ok {
+			depth = ctxDepth
+		}
+		if int(file.Size()) > maxSize {
+			lCtx.Logger().V(3).Info("skipping file due to size")
+			return nil
+		}
+
+		if common.SkipFile(file.Name()) {
+			lCtx.Logger().V(5).Info("skipping file")
+			return nil
+		}
+
+		if h.skipBinaries && common.IsBinary(file.Name()) {
+			lCtx.Logger().V(5).Info("skipping binary file")
+			return nil
+		}
+
+		fReader, err := file.Open()
+		if err != nil {
+			return err
+		}
+		defer fReader.Close()
+
+		return h.openArchive(lCtx, depth, fReader, archiveChan)
+	}
+}
+
+// handleNonArchiveContent processes files that do not contain nested archives, serving as the final stage in the
+// extraction/decompression process. It reads the content to detect its MIME type and decides whether to skip based
+// on the type, particularly for binary files. It manages reading file chunks and writing them to the archive channel,
+// effectively collecting the final bytes for further processing. This function is a key component in ensuring that all
+// file content, regardless of being an archive or not, is handled appropriately.
+func (h *DefaultHandler) handleNonArchiveContent(ctx logContext.Context, reader io.Reader, archiveChan chan []byte) error {
+	bufReader := bufio.NewReaderSize(reader, defaultBufferSize)
+	// A buffer of 512 bytes is used since many file formats store their magic numbers within the first 512 bytes.
+	// If fewer bytes are read, MIME type detection may still succeed.
```

Comment:
Can you use `determineMimeType` here?

Reply:
I considered it, but given this was a lift and shift job, I opted to leave it as is.